File: duplicateFinder/src/main/scala/com/github/michalsenkyr/courses/scala/duplicatefinder/scala/DuplicateFinder.scala
package com.github.michalsenkyr.courses.scala.duplicatefinder.scala

import java.io.{File, FileInputStream}
import java.security.MessageDigest

/**
  * Duplicate file finder in Scala using JRE pre-8
  */
object DuplicateFinder extends App {
  lazy val digest = MessageDigest.getInstance("SHA-1")

  println(args.view
    .map(new File(_))
    .flatMap(listAllFiles)
    .findDuplicates(_.length)
    .flatMap(_.findDuplicates(hash))
    .map(_.mkString(", "))
    .mkString("Duplicate files:\n", "\n", ""))

  def hash(file: File) = digest.digest(file.readBytes()).toSeq

  def listAllFiles(dir: File): Seq[File] = dir.listFiles().flatMap {
    case file if file.isDirectory => listAllFiles(file)
    case file if file.isFile => file :: Nil
    case file => throw new RuntimeException(s"Unknown file type: $file")
  }

  implicit class DuplicateFiltering[T](iterable: Iterable[T]) {
    def findDuplicates[K](f: T => K) = iterable
      .groupBy(f)
      .values
      .filter(_.size > 1)
  }

  implicit class FileReading(file: File) {
    val bufferSize = 4096

    def readBytes() = doWith(new FileInputStream(file)) { input =>
      val buffer = Array.ofDim[Byte](bufferSize)
      Stream.continually {
        buffer.view.take(input.read(buffer))
      }.takeWhile(_.nonEmpty).flatten.toArray
    }
  }

  def doWith[T <: AutoCloseable, R](t: T)(f: T => R) = try f(t) finally t.close()
}

File: examples/duplicateFinder/src/main/java/com/github/michalsenkyr/courses/scala/duplicatefinder/java/DuplicateFinder.java
package com.github.michalsenkyr.courses.scala.duplicatefinder.java;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.*;

/**
 * Duplicate file finder in Java pre-8
 */
public class DuplicateFinder {
    private static MessageDigest digest;

    public static void main(String[] args) {
        // Group by size
        Map<Long, List<File>> bySize = new HashMap<>();
        for (String arg : args) {
            File path = new File(arg);
            for (File file : listAllFiles(path)) {
                long size = file.length();
                List<File> files = bySize.get(size);
                if (files == null) {
                    files = new LinkedList<>();
                    bySize.put(size, files);
                }
                files.add(file);
            }
        }

        // Group by hash
        Map<String, List<File>> byHash = new HashMap<>();
        for (List<File> files : bySize.values()) {
            if (files.size() > 1) {
                for (File file : files) {
                    String hash;
                    try {
                        hash = hash(file);
                    } catch (NoSuchAlgorithmException | IOException e) {
                        throw new RuntimeException(e);
                    }
                    List<File> hashList = byHash.get(hash);
                    if (hashList == null) {
                        hashList = new LinkedList<>();
                        byHash.put(hash, hashList);
                    }
                    hashList.add(file);
                }
            }
        }

        // Print output
        System.out.println("Duplicate files:");
        for (List<File> files : byHash.values()) {
            if (files.size() > 1) {
                for (Iterator<File> it = files.iterator(); it.hasNext(); ) {
                    File file = it.next();
                    if (it.hasNext()) {
                        System.out.print(file.getPath() + ", ");
                    } else {
                        System.out.println(file.getPath());
                    }
                }
            }
        }
    }

    private static List<File> listAllFiles(File dir) {
        List<File> files = new LinkedList<>();
        for (File file : dir.listFiles()) {
            if (file.isDirectory()) {
                files.addAll(listAllFiles(file));
            } else if (file.isFile()) {
                files.add(file);
            } else {
                throw new RuntimeException("Unknown file type: " + file);
            }
        }
        return files;
    }

    private static String hash(File file) throws NoSuchAlgorithmException, IOException {
        if (digest == null) digest = MessageDigest.getInstance("SHA-1");
        return new String(digest.digest(readBytes(file)));
    }

    private static final int BUFFER_SIZE = 4096;

    private static byte[] readBytes(File file) throws IOException {
        List<Byte> bytes = new LinkedList<>();
        try (FileInputStream input = new FileInputStream(file)) {
            byte[] buffer = new byte[BUFFER_SIZE];
            int read;
            while ((read = input.read(buffer)) != -1) {
                for (int i = 0; i < read; ++i) {
                    bytes.add(buffer[i]);
                }
            }
        }
        byte[] result = new byte[bytes.size()];
        for (int i = 0; i < result.length; ++i) {
            result[i] = bytes.get(i);
        }
        return result;
    }
}
